### PR TITLE
cli: Optimize container research

### LINKED ISSUE #212: virtcontainers: Improve container searching
Currently, the more we are running some containers on the system, the more time it takes our runtime to start new ones. The reason behind is that we always ask the list of all sandboxes in order to parse it from the CLI. This parsing resulting in finding if a container exists or not, and in case it does exist, we need its associated sandbox ID.

An easy way to simplify the current complexity O(n²) into an O(1), would be to use a hash table listing all the containers as a key and their associated sandbox as a value.

We should keep track of the containers and their associated pods through a hash table stored on the host filesystem, in order to support the short lifetime of the CLI process.

### PR DESCRIPTION
**Pull request by sboeuf**
This commit will allow for better performance regarding the time spent
to retrieve the sandbox ID related to a container ID.

The way it works is by relying on a specific mapping between container
IDs and sanbox IDs, meaning it allows to retrieve directly the sandbox
ID related to a container ID from the CLI. This lowers complexity from
O(n²) to O(1), because we don't need to call into ListPod() which was
parsing all the pods and all the containers on the system everytime
the CLI need to retrieve this mapping.

This commit also updates the whole unit tests as a consequence. This
is involving most of them since they were all relying on ListPod()
before.

Fixes #212

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

### CODE REVIEW

**Comment by sboeuf** [comment]
This PR is the same as https://github.com/kata-containers/runtime/pull/213, but it implements the container<==>sandbox mapping from the CLI.
@egernst @jodh-intel @bergwolf @grahamwhaley @amshinde PTAL

File: cli/oci.go [hunk 1]
```diff
@@ -40,30 +42,32 @@ var cgroupsDirPath string
 
 var procMountInfo = "/proc/self/mountinfo"
 
+var ctrsMapTreePath = "/var/run/kata-containers/containers-mapping"
```
Comment (bergwolf) [inline cli/oci.go]: Please put it in the runtime root storage directory instead of a hard coded one.

Comment (sboeuf) [reply cli/oci.go]: As agreed, this should be done in a follow up PR introducing the support of `kata-runtime --root`

**Comment by jodh-intel** [comment]
Although this is now mergeable, can we finish the discussion we're having on #213 as both these PRs are trying to solve the same problem?

Marking as dnm for now...

**Comment by jodh-intel** [comment]
@sboeuf - please can you change the commit message to explain *how* you've optimised the search ;)

**Comment by sboeuf** [comment]
@jodh-intel commit message updated.

File: cli/oci.go [hunk 2]
```diff
@@ -339,3 +343,63 @@ func getCgroupsDirPath(mountInfoFile string) (string, error) {
...
+// ID directory. If there are several files, we could not determine which
+// file name corresponds to the sandbox ID associated, and this would throw
+// an error.
+func fetchContainerIDMapping(containerID string) (string, error) {
+	if containerID == "" {
+		return "", fmt.Errorf("Missing container ID")
+	}
+
+	dirPath := filepath.Join(ctrsMapTreePath, containerID)
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
```
Comment (amshinde) [inline cli/oci.go]: You can get rid of this check. The ReadDir call below will return an error if the directory does not exist.

File: cli/create.go [hunk 3]
```diff
@@ -239,6 +239,10 @@ func createSandbox(ociSpec oci.CompatOCISpec, runtimeConfig oci.RuntimeConfig,
 		return vc.Process{}, fmt.Errorf("BUG: Container list from sandbox is wrong, expecting only one container, found %d containers", len(containers))
 	}
 
+	if err := addContainerIDMapping(containerID, sandbox.ID()); err != nil {
```
Comment (amshinde) [inline cli/create.go]: I think we talked that this could be done in virtcontainers itself. Any reason this is being done in the CLI?

Comment (sboeuf) [reply cli/create.go]: Well I was about to say that we were still needing a lock, but now I am confused. I think this could work:
- `fetchSandboxIDs(containerID)` returns a list of what is inside the directory `/.../containerID/`. No lock needed since the call into `readdir()` is atomic.
- `addSandboxID(containerID, sandboxID)` would add a new sandbox using `os,MkdirAll()`, meaning the path `/.../containerID/sandboxID` would be created, no matter if the parent directory `/.../containerID` exist or not.
- `delSandboxID(container, sandboxID)` would remove the path `/.../containerID/sandboxID` and after, it would call into `os.Rmdir()` `/.../containerID` without checking the error. Indeed, the call to rmdir might fail if something is still in the directory, and also if a `addSandboxID()` has just been called.

Comment (sboeuf) [reply cli/create.go]: Ok so from our offline discussion, `os.MkdirAll()` is obviously not atomic, and that's the reason why we would need some locking. But given the fact that this case would never happen since we will not use it for real, we will never hit it. Let me ask others about what they think.

**Comment by sboeuf** [comment]
@bergwolf @jodh-intel @egernst There is one alternative solution which consists in not being completely bug free from a virtcontainers perspective (a potential container directory creation could happen at the same time another sandboxID is removed, and only in case the container directory becomes empty), but assuming that the corner case has a very low probability in theory, and given that it won't happen at all for us since only the CLI will use it.
I don't know if that's something we can consider in order to make sure we let this implementation inside virtcontainers, and this way we don't spread the storage.
So to summarize, we have 0% chance of hitting this issue, but from virtcontainers perspective, this is not completely safe. We might want to add a retry for the `addSandboxID()` in order to reduce further the chance of failure.

**Comment by amshinde** [comment]
@sboeuf Like I mentioned in your other PR for CLI, I think this could well be implemented inside virtcontainers itself, ie virtcontainers handling the creation/deletion of  sandox entries inside a container directory.

There is a slight chance of a race when the sandbox directory entry is created during container creation since this involves creating/checking the container directory followed by creating the sandbox directory itself(/run/virtcontainers/{cid}/{sandbox_id}). The container directory may be deleted by another process between the two steps. Assuming that we run into this issue, we can retry creating the sandbox entry a few times. (Assuming processes context switch between the two steps exactly !)

We may run into this issue based on the supposition that containers may end up having the same container-id. Now, our cli and docker itself makes sure that container-ids are unique. Since CRIO relies on OCI, container-ids would be unique in CRIO case as well.
@bergwolf I think this is true for frakti as well?

So I think we could just go with the simpler approach and avoid all the locking.

**Comment by bergwolf** [comment]
@amshinde frakti would rely on virtcontainers to ensure the uniqueness of sandboxID globally, and the uniqueness of containerID within a sandbox. For kata cli, it verifies the uniqueness of contaienrID itself but it is really unnecessary.

@sboeuf I'm fine with implementing it inside virtcontainers. I just want to make sure that we use a lockless solution and do not rely on a global map file.

**Comment by sboeuf** [comment]
@bergwolf ok then I'll wait for @jodh-intel @grahamwhaley and @egernst input on that.

**Comment by sboeuf** [comment]
@egernst @bergwolf could you please merge it ?

File: cli/oci.go (hunk 2 above)
Comment (bergwolf) [reply cli/oci.go]: Yes, indeed.

**Comment by sboeuf** [comment]
@bergwolf comment addressed.
